### PR TITLE
[ Conv2D ] Draft : backwarding

### PR DESCRIPTION
**Pull request by jijoongmoon**
This PR provides backwarding of Convolution 2D Layer.

( This PR does not include evaluation of implementation. 
 Still looking for it and if it is found, unit test with evaluation will be ready.)

**Self evaluation:**
1. Build test:	 [X]Passed [ ]Failed [ ]Skipped
2. Run test:	 [X]Passed [ ]Failed [ ]Skipped

Signed-off-by: jijoong.moon <jijoong.moon@samsung.com>

### CODE REVIEW

File: nntrainer/include/conv2d_layer.h [hunk 1]
```diff
@@ -144,14 +144,18 @@ class Conv2DLayer : public Layer {
...
-   * @param[in] kernel convolution kernel
+   * @param[in] in input tensor data
+   * @param[in] indim input tensor dimension
+   * @param[in] kernel convolution kernel data
+   * @param[in] kdim convolution kernel dimension
+   * @param[in] out output
    * @param[in] stride stride value : x, y direction
-   * @param[out] status output of status
-   * @retval Tensor outoput tensor
+   * @param[out] bias bias data
```
Comment (dongju-chae) [inline nntrainer/include/conv2d_layer.h:153]: out -> in

File: nntrainer/src/conv2d_layer.cpp [hunk 2]
```diff
@@ -278,18 +362,20 @@ Tensor Conv2DLayer::conv2d(Tensor in, Tensor kernel, unsigned int const *stride,
...
-            sum +=
-              kernel.getValue(0, i, ki, kj) * in.getValue(0, i, j + ki, k + kj);
+            sum += kernel[i * k_height * k_width + ki * k_width + kj] *
+                   in[i * height * width + (j + ki) * width + (k + kj)];
           }
         }
       }
-      output.setValue(0, 0, I, J, sum);
+      if (bias != 0.0)
+        sum += bias;
```
Comment (dongju-chae) [inline nntrainer/src/conv2d_layer.cpp:371]: Maybe unnecessary if-statement cuz it's 0?

File: test/unittest/unittest_nntrainer_layers.cpp [hunk 3]
```diff
@@ -572,6 +572,54 @@ TEST(nntrainer_Conv2DLayer, forwarding_02_p) {
...
+  EXPECT_EQ(status, ML_ERROR_NONE);
+  layer.setInputDimension(previous_dim);
+
+  status = layer.initialize(true);
+  EXPECT_EQ(status, ML_ERROR_NONE);
+
+  nntrainer::Tensor in(1, 3, 7, 7);
+  nntrainer::Tensor out;
+  nntrainer::Tensor derivatives(1, 2, 5, 5);
+  float golden[50] = {0.25, 0.5, 0.5, 0.5, 0.25, 0.5,  1.0, 1.0, 1.0, 0.5,
```
Comment (kparichay) [inline test/unittest/unittest_nntrainer_layers.cpp:606]: can we rename this from `golden` to `sample_derivative`. 
Golden sounds like derivative is going to be matched with it, but it is just a sample value to check that code runs smoothly.

Comment (jijoongmoon) [reply test/unittest/unittest_nntrainer_layers.cpp]: fixed. thanks.

File: nntrainer/src/conv2d_layer.cpp (hunk 2 above)
Comment (jijoongmoon) [reply nntrainer/src/conv2d_layer.cpp]: fixed. thanks.

File: nntrainer/include/conv2d_layer.h (hunk 1 above)
Comment (jijoongmoon) [reply nntrainer/include/conv2d_layer.h]: fixed. thanks.

**Pull request merged**